### PR TITLE
Interactive interface with autocompletion that can help to select the packages to update

### PR DESCRIPTION
**Pull request by SofHad**
Here is an illustration
en2

### CODE REVIEW

**Comment by hkdobrev** [comment]
@SofHad 

1. You can assume by default the answer to "Add more packages" is "yes" and then just have a sentence on top which says "Empty package ends submission" (or something better). When the user submits empty package name, Composer would assume all packages are entered.

2. This should actually be a flag to the `composer require` command. `composer update` could have interaction to specify just a subset of packages to update interactively, but not add new packages.

3. I don't see myself using this though. The `composer require` command is quite sufficient for my use. Could you provide some examples where a similar feature exists in another dependency/package manager?

**Comment by SofHad** [comment]
Hi @hkdobrev and thank you for your feedbacks.
The idea is to provide the highest degree of user comfort, I think it is suited to composer update.
We can adapt it to the composer require command too.
I try to make the adjustments that you have suggested as soon as i can.

**Comment by SofHad** [comment]
This is done

en3

File: src/Composer/Command/UpdateCommand.php [hunk 1]
```diff
@@ -135,4 +145,102 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
         return $install->run();
     }
+
+    private function getPackagesInteractively(
+        InputInterface $input,
+        OutputInterface $output,
+        Composer $composer
```
Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: you should keep this on 1 line

File: src/Composer/Command/UpdateCommand.php [hunk 2]
```diff
@@ -135,4 +145,102 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+    ) {
+        if ($input->getArgument('packages')) {
+            throw new \InvalidArgumentException(
+                'The option --interactive does not require an argument'
```
Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: why forbidding to use them together ? Just treat them as already included

File: src/Composer/Command/UpdateCommand.php [hunk 3]
```diff
@@ -135,4 +145,102 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+            array_merge($InstalledPackageNames, $requiredPackageNames)
+        );
+
+        $vendorDir      = $config->get('vendor-dir', true);
+        $vendorWildcard = sprintf('%s/*', $vendorDir);
+
+        $autocompleterValues = array($vendorWildcard);
+        foreach ($names as $name) {
+            $autocompleterValues[] = $name;
+            $autocompleterValues[] = sprintf('%s/%s', $vendorDir, $name);
```
Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: I don't understand the reason to pass a vendor directory to an autocompleter for package names

File: src/Composer/Command/UpdateCommand.php (hunk 1 above)
Comment (SofHad) [inline src/Composer/Command/UpdateCommand.php]: I will do it.

File: src/Composer/Command/UpdateCommand.php (hunk 2 above)
Comment (SofHad) [inline src/Composer/Command/UpdateCommand.php]: I think it would be a good idea.
Ok, I will do it too.

**Comment by SofHad** [comment]
To have the autocomplete suggestions for the words starting with vendor:
* vendor/*
* vendor/doctrine/orm
* doctrine/orm
* why not for vendor/doctrine/*

what do you think?

**Comment by stof** [comment]
``vendor/doctrine/*`` does not make sense. The package whitelist is about package names. ``vendor/doctrine/orm`` and ``vendor/doctrine/*`` are invalid choices

**Comment by SofHad** [comment]
vendor/doctrine/*  was not yet implemented contrary to vendor/doctrine/orm.
 I will update it. I leave only:
* vendor/*
* doctrine/orm

**Comment by SofHad** [comment]
@stof 
I fixed these elements and removed the vendor for the autocomplete values.

File: src/Composer/Command/UpdateCommand.php [hunk 4]
```diff
@@ -74,13 +79,17 @@ protected function configure()
 
     protected function execute(InputInterface $input, OutputInterface $output)
     {
+        $composer = $this->getComposer(true, $input->getOption('no-plugins'));
```
Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: This is wrong. You cannot move it before the handling of the ``--no-custom-installers`` option

Comment (SofHad) [inline src/Composer/Command/UpdateCommand.php]: Yes, of course. 
This is done :)

File: src/Composer/Command/UpdateCommand.php [hunk 5]
```diff
@@ -78,9 +83,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
             $output->writeln('<warning>You are using the deprecated option "no-custom-installers". Use "no-plugins" instead.</warning>');
             $input->setOption('no-plugins', true);
         }
-
         $composer = $this->getComposer(true, $input->getOption('no-plugins'));
-        $composer->getDownloadManager()->setOutputProgress(!$input->getOption('no-progress'));
```
Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: removing this line is wrong

File: src/Composer/Command/UpdateCommand.php [hunk 6]
```diff
@@ -78,9 +83,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+
+        $packages = $input->getArgument('packages');
+        if ($input->getOption('interactive')) {
```
Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: use ``$input->isInteractive()`` instead for the proper way to check for interactive input

File: src/Composer/Command/UpdateCommand.php [hunk 7]
```diff
@@ -78,9 +83,13 @@ protected function execute(InputInterface $input, OutputInterface $output)
             $output->writeln('<warning>You are using the deprecated option "no-custom-installers". Use "no-plugins" instead.</warning>');
             $input->setOption('no-plugins', true);
         }
-
```
Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: please keep this line

File: src/Composer/Command/UpdateCommand.php [hunk 8]
```diff
@@ -135,4 +144,80 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+
+            if (is_string($addedPackage)) {
+                if (!in_array($addedPackage, $packages)) {
+                    $packages[] = $addedPackage;
+                    $autocompleterValues = array_diff($autocompleterValues, array($addedPackage));
+                } else {
+                    $output->writeln(sprintf('<error>The package "%s" was already added.</error>', $package));
+                }
+            } else {
+                $continue = false;
```
Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: use ``break;`` to break the loop instead of this variable

Comment (SofHad) [inline src/Composer/Command/UpdateCommand.php]: Done

File: src/Composer/Command/UpdateCommand.php (hunk 5 above)
Comment (SofHad) [inline src/Composer/Command/UpdateCommand.php]: I don't know how it happened. :grimacing:

File: src/Composer/Command/UpdateCommand.php (hunk 6 above)
Comment (SofHad) [inline src/Composer/Command/UpdateCommand.php]: I added

```php
private function isInteractive($input)
{
     return $input->getOption('interactive');
}
```

Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: hmm, sorry, I confused the meaning of the ``interactive`` option here. No need to add a getter for that

File: src/Composer/Command/UpdateCommand.php [hunk 9]
```diff
@@ -135,4 +147,84 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
         return $install->run();
     }
+
+    private function getPackagesInteractively(InputInterface $input, OutputInterface $output, Composer $composer, $packages)
+    {
```
Comment (stof) [inline src/Composer/Command/UpdateCommand.php]: you should do a check on ``$input->isInteractive()`` here and throw an exception saying that ``--interactive`` cannot be used in non-interactive terminals

Comment (SofHad) [inline src/Composer/Command/UpdateCommand.php]: Done

**Comment by SofHad** [comment]
ping @naderman @Seldaek

**Comment by Seldaek** [comment]
Thanks @SofHad and sorry for the delay looking at this

**Comment by SofHad** [comment]
@stof thank you for your cooperation ;-)

File: src/Composer/Command/UpdateCommand.php [hunk 10]
```diff
@@ -135,4 +147,83 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+        $requiredPackageNames = array();
+        foreach (
+            array_merge(
+                $composer->getPackage()->getRequires(),
+                $composer->getPackage()->getDevRequires()
+            ) as $require) {
+            $requiredPackageNames[] = $require->getTarget();
+        }
+
+        $InstalledPackageNames = array();
```
Comment (SofHad) [inline src/Composer/Command/UpdateCommand.php]: I will fix this PSR error in another PR

**Comment by Seldaek** [comment]
@SofHad I already fixed the PSR stuff in 633e9f91a1a93bec5a22843cf1bba7d2bf97260a don't worry about it.

**Pull request merged**